File: 0305/SeoyeonHong/2644 촌수 계산.py
import sys
from collections import deque

input = sys.stdin.readline
n = int(input())
a, b = map(int, input().split())
m = int(input())
graph = [[] for _ in range(n+1)]
visited = [False for _ in range(n+1)]
relative = False
for _ in range(m):
    x, y = map(int, input().split())
    graph[x].append(y)
    graph[y].append(x)

q = deque([(a, 0)])
visited[a] = True

while q:
    cur, cnt = q.popleft()
    if cur == b:
        print(cnt)
        relative = True
        break
    for next in graph[cur]:
        if not visited[next]:
            visited[next] = True
            q.append((next, cnt+1))

if not relative:
    print(-1)
    

File: 0305/SoohyunKim/2644_촌수계산.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const N = +input[0];
const [p1, p2] = input[1].split(' ').map(Number);
const M = +input[2];

let visited = Array(N + 1).fill(false);

let graph = Array.from(Array(N + 1), () => []);
for(let i = 0; i < M; i++) {
  const [from, to] = input[i+3].split(' ').map(Number);
  graph[from].push(to);
  graph[to].push(from);
}

function dfs(x, y, cnt) {
  if(x === y) return cnt;

  visited[x] = true;
  for(let i = 0; i < graph[x].length; i++) {
    const next = graph[x][i];
    
    if(!visited[next]){
      const ans = dfs(next, y, cnt+1);
      if(ans!== -1) return ans;
    }
  }

  return -1
} 

console.log(dfs(p1, p2, 0));